feat: разработать согласно ТЗ-14 ShareIt

aigeley opened the pull request.

aigeley committed:
feat: разработать согласно ТЗ-14 ShareIt

eugene-lenkevich commented on src/main/java/ru/practicum/shareit/booking/model/BookingMapper.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public static Booking toBooking(Booking booking, BookingDto bookingDto, Item item, User booker, BookingStatus status) {
+        Optional.ofNullable(bookingDto.getId()).ifPresent(booking::setId);
```
Можно наверно было сделать и без Optional, NullPointerException не упала бы если бы поля были null, преобразований никаких же нет :)

aigeley replied:
Исправлено, оставил Optional только там, где может быть частичное обновление полей через PATCH.

eugene-lenkevich commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,126 @@
+                break;
+        }
+
+        Specification<Booking> specification = Specification
```
Это API достаточно тяжелое для понимания, для простых случаев (методы выше) лучше его избегать, а вот тут вполне себе оправдано -  некоторое конструирование запросов в зависити от состояния :)

aigeley replied:
Переписал ItemRepository::search на JPQL.
BookingRepositoryCustomImpl пока оставил как было, многие объявленные условия переиспользую повторно - могу ли в нём тоже как-то сократить запись?

eugene-lenkevich replied:
А вот там как раз пусть остается, там как раз используется повторно. Осоьо короче не уверен, главное чтобы понятно было что куда/

А вот сократить можно еще вот как. Паттерн "стратегия".

Если есть желание можешь попробовать применить паттерн "стратегия"

Возможно, именно для этого случая стратегия покажется избыточной (обработчики все в одну строчку), но часто в обработчике бывает нетривиальная логика и куча своих зависимостей и там это оправдано :) 

Смысл такой. У задаем соответствие между состоянием и обработчиком этого состояния.
Можно сделать мапу
Map<String, BookingStateFetchStrategy>, где BookingStateFetchStrategy - который будут реализовывать обработчики состояния

или же, более простой вариант, в мапу засунуть методы сервиса, примерно так будет
Map<String, Supplier> handlers = ...
handlers.put("ALL", this::handleAllState) 

То есть, мы заменяем switch на поиск нужного обработчика через map :)
Это немного не про сами запросы, а про более общую историю


https://ravthiru.medium.com/strategy-design-pattern-with-in-spring-boot-application-2ff5a7486cd8

aigeley replied:
Попробую переписать в следующем спринте, спасибо!

eugene-lenkevich commented on src/main/java/ru/practicum/shareit/item/model/CommentMapper.java:
```diff
@@ -0,0 +1,36 @@
+        return comment == null ? null : new CommentDto(
+                comment.getId(),
+                comment.getText(),
+                comment.getItem().getId(),
```
Вот как раз тут и нужен Optional :)

Optiona.ofNullable(comment.getItem).map(Item::getId).orElse(null)

aigeley replied:
Исправлено, добавил Optional везде, где запрашивается поле вложенного объекта.

eugene-lenkevich commented on src/main/java/ru/practicum/shareit/user/model/UserMapper.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public static User toUser(User user, UserDto userDto) {
+        Optional.ofNullable(userDto.getId()).ifPresent(user::setId);
```
Я заметил в мапперах везде такая логика используется. В данном случае идет обновление не-null полей, так и положенно, то есть если поле в исходном объектое стало null, то оно не обновится в объекте-назначении(если там что то уже есть), это нормальное поведение? :)

aigeley replied:
Можем ли как-то в PATCH определить, что поле не было передано совсем (т.е. не нужно его обновлять) или что специально было передано значение field: null (нужно обновить поле на null)?
Так понял, выборочно обновить поля на null для существующего объекта можно, если заново передать объект целиком через PUT, но в ТЗ такого метода пока нет...

eugene-lenkevich replied:
Да не, не надо так заморачиваться, чтобы такое сделать потребуется переделать модель данных - пересылать непосрдественно список полей и уже их менять. У нас же приходит сущность, ее и обновляем, не думаю что тут стоит сильно усложнять логику, в большинстве случаев такого подхода достаточно :)

А еще, можно сделать специальнй метод для обновления принимающий на вход не всю сущность а некое dto  с ограниченным набором полей. Таким образом обновляться будут только те поля, которые есть в dto

eugene-lenkevich reviewed (changes requested):
Привет, Андрей, спасибо за PR

В целом все отлично, все сущости описаны как надо, единственное я оставил нескольк вопросов по испоьзованию Optional, посмотри комментарии, пожалуйста :)

aigeley committed:
feat: исправить замечания ревью

- оставить Optional только там, где может быть частичное обновление полей через PATCH
- добавить Optional везде, где запрашивается поле вложенного объекта
- переписать ItemRepository::search на JPQL

eugene-lenkevich reviewed (approved):
Теперь все отлично, спасибо :)
